### PR TITLE
Configure service levels interval

### PR DESCRIPTION
**Pull request by Jadw1**
Service level controller updates itself in interval. However the interval time is hardcoded in main to 10 seconds and it leads to long sleeps in some of the tests.

This patch moves this value to `service_levels_interval_ms` command line option and sets this value to 0.5s in cql-pytest.

### CODE REVIEW

File: test/cql-pytest/suite.yaml
Comment (nyh) [inline test/cql-pytest/suite.yaml:6]: Any particular reason why this should be in test/cql-pytest/suite.yaml and not in test/pylib/scylla_cluster.py with all the rest of the options that we use to configure Scylla? As you can see test/cql-pytest/suite.yaml didn't yet list any options, and I don't think we should start -  I think it's confusing that each option appears in a different file and nobody knows where to find the relevant options.

Please also add the same option to test/cql-pytest/run.py which has another selection of options which are used when running test/cql-pytest/run or test/alternator/run. Thanks.

**Review by nyh — CHANGES_REQUESTED** [review]

**Review by michoecho — CHANGES_REQUESTED** [review]
Make it live-updatable, please. If a piece config can be live-updateable (and this one very much can), it should be.

File: main.cc
Comment (michoecho) [inline main.cc:1147]: Nit: this is fine, but it would be better if this function returned a `std::chrono::duration` instead of `uint32_t` milliseconds.

Ideally only the config layer should care/know about the time unit. It needs that, because our only way to specify some duration in the config is by typing an integer in some agreed unit.

But other layers should communicate with `duration`s, because this removes the risk of confusing the unit somewhere down the line. (And it frees you from the need to document the unit).

File: test/cql-pytest/suite.yaml
Comment (eliransin) [inline test/cql-pytest/suite.yaml:4]: It looks like this change is unintentional

**Review by eliransin — APPROVED** [review]
Please remove the unrelated change in `test/cql-pytest/suite.yaml`
I also agree with @michoecho about converting the units in the configuration layer.
So I am assuming those two things will be fixed before merge :slightly_smiling_face:

File: test/cql-pytest/suite.yaml
Comment (Jadw1) [reply test/cql-pytest/suite.yaml]: dropped

**Comment by Jadw1** [comment]
I've changed the returned value, so `service_levels_interval` returns `steady_clock_type::duration` but I'm missing sth and currently the PR won't compile. @michoecho can you see what I'm missing? From the compilation error message it looks like sth with operator<< and operator>> but I can't find what exactly.

**Comment by mykaul** [comment]
```
00:07:32.518  db/config.cc:366:13: note: in instantiation of member function 'utils::config_file::named_value<std::chrono::duration<long, std::ratio<1, 1000000000>>>::add_command_line_option' requested here
00:07:32.519  db::config::config(std::shared_ptr<db::extensions> exts)
00:07:32.519              ^
00:07:32.519  /usr/include/boost/lexical_cast/detail/converter_lexical.hpp:243:54: note: expression evaluates to 'false || false'
00:07:32.519              BOOST_STATIC_ASSERT_MSG((result_t::value || boost::has_right_shift<std::basic_istream<wchar_t>, T >::value),
00:07:32.519                                       ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
00:07:32.519  /usr/include/boost/static_assert.hpp:32:59: note: expanded from macro 'BOOST_STATIC_ASSERT_MSG'
00:07:32.519  #     define BOOST_STATIC_ASSERT_MSG( ... ) static_assert(__VA_ARGS__)
00:07:32.519                                                            ^~~~~~~~~~~
00:07:32.519  In file included from db/config.cc:16:
```

Build failure.

**Comment by michoecho** [comment]
> I've changed the returned value, so `service_levels_interval` returns `steady_clock_type::duration` but I'm missing sth and currently the PR won't compile. @michoecho can you see what I'm missing? From the compilation error message it looks like sth with operator<< and operator>> but I can't find what exactly.
> 
> In the patch I've put `operator<<(std::ostream& os, const steady_clock_type::duration&)` and `operator>>(std::istream& is, steady_clock_type::duration&)` in db namespace, which I think is not correct. But when I moved the functions out of it simply to config.hh, it also doesn't work.

@Jadw1 Config code calls `boost::lexical_cast` on `steady_clock_type::duration&`, which calls `operator>>(std::istream&, steady_clock_type::duration&)`.
But due to the rules of the language (https://en.cppreference.com/w/cpp/language/adl), it only looks for that function in namespaces `boost` (that's the call site) and `std` (because the function has arguments from that namespace). And the function isn't defined there, but in namespace `db`.

To solve this, you can either define the operator in namespace `std` or `boost` (which would be technically illegal, so let's avoid it), or you can wrap the duration into some type defined in namespace `db`. For example, this works:
```diff
diff --git a/db/config.cc b/db/config.cc
index e1ab60a364..c9a7e73db6 100644
--- a/db/config.cc
+++ b/db/config.cc
@@ -74,7 +74,7 @@ hinted_handoff_enabled_to_json(const db::config::hinted_handoff_enabled_type& h)
 static 
 json::json_return_type
 duration_to_json(const db::config::duration_type& d) {
-    return value_to_json(std::chrono::duration_cast<std::chrono::milliseconds>(d).count());
+    return value_to_json(std::chrono::duration_cast<std::chrono::milliseconds>(d.val).count());
 }
 
 // Convert a value that can be printed with operator<<, or a vector of
@@ -308,7 +308,7 @@ struct convert<db::config::duration_type> {
         if (!convert<int64_t>::decode(node, opt)) {
             return false;
         }
-        rhs = std::chrono::duration_cast<db::config::duration_type>(std::chrono::milliseconds(opt));
+        rhs.val = std::chrono::milliseconds(opt);
         return true;
     }
 };
@@ -1148,7 +1148,7 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , maximum_replication_factor_fail_threshold(this, "maximum_replication_factor_fail_threshold", liveness::LiveUpdate, value_status::Used, -1, "")
     , replication_strategy_warn_list(this, "replication_strategy_warn_list", liveness::LiveUpdate, value_status::Used, {locator::replication_strategy_type::simple}, "Controls which replication strategies to warn about when creating/altering a keyspace. Doesn't affect the pre-existing keyspaces.")
     , replication_strategy_fail_list(this, "replication_strategy_fail_list", liveness::LiveUpdate, value_status::Used, {}, "Controls which replication strategies are disallowed to be used when creating/altering a keyspace. Doesn't affect the pre-existing keyspaces.")
-    , service_levels_interval(this, "service_levels_interval_ms", liveness::LiveUpdate, value_status::Used, std::chrono::duration_cast<db::config::duration_type>(std::chrono::milliseconds(10000)),
+    , service_levels_interval(this, "service_levels_interval_ms", liveness::LiveUpdate, value_status::Used, {std::chrono::milliseconds(10000)},
         "Controls how often service levels module polls configuration table")
     , error_injections_at_startup(this, "error_injections_at_startup", error_injection_value_status, {}, "List of error injections that should be enabled on startup.")
     , default_log_level(this, "default_log_level", value_status::Used)
@@ -1239,14 +1239,14 @@ std::istream& operator>>(std::istream& is, error_injection_at_startup& eias) {
     return is;
 }
 
-std::ostream& operator<<(std::ostream& os, const steady_clock_type::duration& d) {
-    os << std::chrono::duration_cast<std::chrono::milliseconds>(d).count();
+std::ostream& operator<<(std::ostream& os, const my_duration_type& d) {
+    os << std::chrono::duration_cast<std::chrono::milliseconds>(d.val).count();
     return os;
 }
-std::istream& operator>>(std::istream& is, steady_clock_type::duration& d) {
+std::istream& operator>>(std::istream& is, my_duration_type& d) {
     int64_t ms;
     is >> ms;
-    d = std::chrono::duration_cast<steady_clock_type::duration>(std::chrono::milliseconds(ms));
+    d.val = std::chrono::duration_cast<steady_clock_type::duration>(std::chrono::milliseconds(ms));
     return is;
 }
 
diff --git a/db/config.hh b/db/config.hh
index 7a8ee04279..79894d13c2 100644
--- a/db/config.hh
+++ b/db/config.hh
@@ -79,11 +79,16 @@ struct error_injection_at_startup {
     }
 };
 
+struct my_duration_type {
+    steady_clock_type::duration val;
+    bool operator==(const my_duration_type&) const = default;
+};
+
 std::ostream& operator<<(std::ostream& os, const error_injection_at_startup&);
 std::istream& operator>>(std::istream& is, error_injection_at_startup&);
 
-std::ostream& operator<<(std::ostream& os, const steady_clock_type::duration&);
-std::istream& operator>>(std::istream& is, steady_clock_type::duration&);
+std::ostream& operator<<(std::ostream& os, const my_duration_type&);
+std::istream& operator>>(std::istream& is, my_duration_type&);
 
 }
 
@@ -165,7 +170,7 @@ class config final : public utils::config_file {
     using seed_provider_type = db::seed_provider_type;
     using hinted_handoff_enabled_type = db::hints::host_filter;
     using error_injection_at_startup = db::error_injection_at_startup;
-    using duration_type = steady_clock_type::duration;
+    using duration_type = db::my_duration_type;
 
     /*
      * All values and documentation taken from
```

**Comment by michoecho** [comment]
@Jadw1 But the latest changes aren't good. You defined config input/output for `steady_clock_type::duration` in terms of milliseconds, but we don't always pass durations in the config as milliseconds — sometimes we pass seconds. Writing a canonical `duration_type` that only works for millisecond-unit entries is awkward.

You went too far. I was only suggesting this:
```diff
           sl_controller.local().update_from_distributed_data([cfg] () {
-              return cfg->service_levels_interval();
+              return std::chrono::milliseconds(cfg->service_levels_interval());
           });
```
I wasn't suggesting to make `service_levels_interval` a `named_value<duration_type>`.

We can take the latter path too (IMHO it would be better than having primitive integers there), but in that case we should add a duration wrapper which can handle many duration types (in particular, integer milliseconds, integer seconds, floating-point seconds) and convert all duration-type config entries to it. It's a topic for a different PR.

**Comment by Jadw1** [comment]
Oh, ok. I see. It should be ok now.

**Comment by Jadw1** [comment]
@nyh @michoecho Ping, all of the comments are addressed

**Review by michoecho — APPROVED** [review]
LGTM

The lower the interval in tests, the more spam in logs there will be in tests where the the `service_levels` query can't be executed temporarily.

But we can fix that if we see it's a problem.

File: service/qos/service_level_controller.hh
Comment (denesb) [inline service/qos/service_level_controller.hh:138]: Why lambda? To support live-update? In that case, why not use updateable_value instead?

Comment (denesb) [reply service/qos/service_level_controller.hh:138]: Well, it does. When you have a specific type for something, it is best to use it, instead of some roughly equivalent generic type. The type used communicates intent. A generic lambda communicates nothing.

Comment (michoecho) [reply service/qos/service_level_controller.hh:138]: In principle I don't disagree, it's just that this 10-liner patch has already been in review for a month.

I'm getting the impression that the shorter a patch is, the stricter the reviews.

Comment (denesb) [reply service/qos/service_level_controller.hh:138]: > In principle I don't disagree, it's just that this 10-liner patch has already been in review for a month.

These are rookie numbers. I have a one liner patch, adding an empty line to the docs, waiting for merge for 2+ months.


> I'm getting the impression that the shorter a patch is, the stricter the reviews.

Sometimes it is the case. Fine, im the name of the good-enough principle, I will queue when I get home, but @Jadw1 please fix in a follow-up.

**Pull request merged**
